### DISCUSSION TITLE
SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.

### DISCUSSION THREAD

**Discussion by KfearO** [MatthiasValvekens/pyHanko#241] [Q&A]
PS C:\Users\kfear\PycharmProjects\DigitallySignPdf> pip install 'pyHanko[opentype]'
Requirement already satisfied: pyHanko[opentype] in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (0.17.2)
Requirement already satisfied: asn1crypto>=1.5.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (1.5.1)
Requirement already satisfied: pytz>=2020.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (2022.7.1)
Requirement already satisfied: qrcode>=6.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (7.4.2)
Requirement already satisfied: tzlocal>=2.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (4.2)
Requirement already satisfied: pyhanko-certvalidator==0.20.* in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (0.20.1)
Requirement already satisfied: click>=7.1.2 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (8.1.3)
Requirement already satisfied: requests>=2.24.0 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (2.28.2)
Requirement already satisfied: pyyaml>=5.3.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (6.0)
Requirement already satisfied: cryptography>=3.3.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (39.0.2)
Collecting fonttools>=4.33.3
  Using cached fonttools-4.39.0-py3-none-any.whl (1.0 MB)
Collecting uharfbuzz<0.31.0,>=0.25.0
  Using cached uharfbuzz-0.30.0.zip (1.3 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: oscrypto>=1.1.0 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyhanko-certvalidator==0.20.*->pyHanko[opentype]) (1.3.0)
Requirement already satisfied: uritools>=3.0.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyhanko-certvalidator==0.20.*->pyHanko[opentype]) (4.0.1)
Requirement already satisfied: colorama in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from click>=7.1.2->pyHanko[opentype]) (0.4.6)
Requirement already satisfied: cffi>=1.12 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from cryptography>=3.3.1->pyHanko[opentype]) (1.15.1)
Requirement already satisfied: typing-extensions in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from qrcode>=6.1->pyHanko[opentype]) (4.5.0)
Requirement already satisfied: pypng in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from qrcode>=6.1->pyHanko[opentype]) (0.20220715.0)
Requirement already satisfied: charset-normalizer<4,>=2 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (3.1.0)      
Requirement already satisfied: idna<4,>=2.5 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (3.4)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (1.26.15)       
Requirement already satisfied: certifi>=2017.4.17 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (2022.12.7)        
Requirement already satisfied: pytz-deprecation-shim in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from tzlocal>=2.1->pyHanko[opentype]) (0.1.0.post0)
Requirement already satisfied: tzdata in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from tzlocal>=2.1->pyHanko[opentype]) (2022.7)
Requirement already satisfied: pycparser in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from cffi>=1.12->cryptography>=3.3.1->pyHanko[opentype]) (2.21)
Building wheels for collected packages: uharfbuzz
  Building wheel for uharfbuzz (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for uharfbuzz (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [13 lines of output]
      C:\Users\kfear\AppData\Local\Temp\pip-build-env-131k1o71\overlay\Lib\site-packages\setuptools\config\setupcfg.py:516: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
        warnings.warn(msg, warning_class)
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-311
      creating build\lib.win-amd64-cpython-311\uharfbuzz
      copying src\uharfbuzz\_version.py -> build\lib.win-amd64-cpython-311\uharfbuzz
      copying src\uharfbuzz\__init__.py -> build\lib.win-amd64-cpython-311\uharfbuzz
      running build_ext
      building 'uharfbuzz._harfbuzz' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for uharfbuzz
Failed to build uharfbuzz
ERROR: Could not build wheels for uharfbuzz, which is required to install pyproject.toml-based projects

**Comment by MatthiasValvekens** [comment]
Hi @KfearO,

The warning you're seeing is unrelated to the actual problem. This is the real issue:

> building 'uharfbuzz._harfbuzz' extension
> error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
> [end of output]

> note: This error originates from a subprocess, and is likely not a problem with pip.
> ERROR: Failed building wheel for uharfbuzz
> Failed to build uharfbuzz

As the error explains, you should update your C++ build toolchain if you want to install uharfbuzz from source.

Incidentally, both the warning and the error pertain to [uharfbuzz](https://github.com/harfbuzz/uharfbuzz), and have nothing to do with pyHanko as such. I can relax the version constraint to allow more recent versions, though. Those seem to have prebuilt py3.11 wheels for Windows.

**Comment by KfearO** [comment]
Hi @MatthiasValvekens 

on the right side you can see what was already installed.
can you please explain what am I missing?

![image](https://user-images.githubusercontent.com/30979079/225420050-72de5a66-e56c-42b7-a322-2188e0b39b9d.png)

**Reply by MatthiasValvekens** [reply]
I have never done any C++ development on Windows, so I can't help you with that.

I did just test that pyHanko will run just fine with uharfbuzz 0.35.0, so if you run `pip install uharfbuzz==0.35.0` before installing pyHanko, you might get a warning, but it should work (version 0.35.0 comes with prebuilt python 3.11 wheels for Windows, so you shouldn't need to build anything).

I just pushed a commit to relax that upper version bound, for what it's worth.

**Comment by KfearO** [comment]
still failing

10x for the effort


PS C:\Users\kfear\PycharmProjects\DigitallySignPdf> pip install uharfbuzz==0.35.0  
Collecting uharfbuzz==0.35.0
  Downloading uharfbuzz-0.35.0-cp311-cp311-win_amd64.whl (923 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 924.0/924.0 kB 3.4 MB/s eta 0:00:00
Installing collected packages: uharfbuzz
Successfully installed uharfbuzz-0.35.0
PS C:\Users\kfear\PycharmProjects\DigitallySignPdf> pip install 'pyHanko[opentype]'  
Requirement already satisfied: pyHanko[opentype] in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (0.17.2)
Requirement already satisfied: asn1crypto>=1.5.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (1.5.1)
Requirement already satisfied: pytz>=2020.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (2022.7.1)
Requirement already satisfied: qrcode>=6.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (7.4.2)
Requirement already satisfied: tzlocal>=2.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (4.2)
Requirement already satisfied: pyhanko-certvalidator==0.20.* in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (0.20.1)
Requirement already satisfied: click>=7.1.2 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (8.1.3)
Requirement already satisfied: requests>=2.24.0 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (2.28.2)
Requirement already satisfied: pyyaml>=5.3.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (6.0)
Requirement already satisfied: cryptography>=3.3.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyHanko[opentype]) (39.0.2)
Collecting fonttools>=4.33.3
  Using cached fonttools-4.39.0-py3-none-any.whl (1.0 MB)
Collecting uharfbuzz<0.31.0,>=0.25.0
  Using cached uharfbuzz-0.30.0.zip (1.3 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: oscrypto>=1.1.0 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyhanko-certvalidator==0.20.*->pyHanko[opentype]) (1.3.0)  
Requirement already satisfied: uritools>=3.0.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from pyhanko-certvalidator==0.20.*->pyHanko[opentype]) (4.0.1)  
Requirement already satisfied: colorama in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from click>=7.1.2->pyHanko[opentype]) (0.4.6)
Requirement already satisfied: cffi>=1.12 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from cryptography>=3.3.1->pyHanko[opentype]) (1.15.1)
Requirement already satisfied: typing-extensions in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from qrcode>=6.1->pyHanko[opentype]) (4.5.0)
Requirement already satisfied: pypng in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from qrcode>=6.1->pyHanko[opentype]) (0.20220715.0)
Requirement already satisfied: charset-normalizer<4,>=2 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (3.1.0)
Requirement already satisfied: idna<4,>=2.5 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (3.4)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (1.26.15)       
Requirement already satisfied: certifi>=2017.4.17 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from requests>=2.24.0->pyHanko[opentype]) (2022.12.7)        
Requirement already satisfied: pytz-deprecation-shim in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from tzlocal>=2.1->pyHanko[opentype]) (0.1.0.post0)
Requirement already satisfied: tzdata in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from tzlocal>=2.1->pyHanko[opentype]) (2022.7)
Requirement already satisfied: pycparser in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (from cffi>=1.12->cryptography>=3.3.1->pyHanko[opentype]) (2.21)
Building wheels for collected packages: uharfbuzz
  Building wheel for uharfbuzz (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for uharfbuzz (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [13 lines of output]
      C:\Users\kfear\AppData\Local\Temp\pip-build-env-mukfav8x\overlay\Lib\site-packages\setuptools\config\setupcfg.py:516: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
        warnings.warn(msg, warning_class)
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-311
      creating build\lib.win-amd64-cpython-311\uharfbuzz
      copying src\uharfbuzz\_version.py -> build\lib.win-amd64-cpython-311\uharfbuzz
      copying src\uharfbuzz\__init__.py -> build\lib.win-amd64-cpython-311\uharfbuzz
      running build_ext
      building 'uharfbuzz._harfbuzz' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for uharfbuzz
Failed to build uharfbuzz
ERROR: Could not build wheels for uharfbuzz, which is required to install pyproject.toml-based projects
PS C:\Users\kfear\PycharmProjects\DigitallySignPdf> pip install uharfbuzz==0.35.0  
Requirement already satisfied: uharfbuzz==0.35.0 in c:\users\kfear\pycharmprojects\digitallysignpdf\venv\lib\site-packages (0.35.0)

**Reply by MatthiasValvekens** [reply]
Sorry, I somehow thought you were installing pyHanko from a git checkout. If not, you'll have to wait for the next release.

Alternatively, can you use Python 3.10 instead? I believe there are prebuilt uharfbuzz wheels for 3.10 going back far enough.

**Reply by KfearO** [reply]
all good, no worries, as its an experimental project I'm doing for my own, I'll wait a bit longer and if by the time I'll get back to it there won't be any version ready with python 3.11, I'll install it with python 3.10
10x again for the effort
